week4/json-print/jasonkuhrt/Main.hs

```diff
@@ -0,0 +1,119 @@
+{- README
+
+-- Code Kata : JSON Print
+
+Define a data type for JSON and print it as JSON (without indentation).
+Don't forget that you should support floating-point numbers and escaping in
+strings.
+-}
+module Main where
+
+import Data.Fixed (mod')
+import Data.Map (Map)
+import qualified Data.Map as Map
+import qualified Data.List as List
+
+
+
+data JSON =
+    Nil
+  | B Bool
+  | S String
+  | N Double
+  | L [JSON]
+  | O (Map String JSON)
+  deriving (Show)
+
+
+
+{- | Stringify a JSON type to a valid JSON string. -}
+stringify :: JSON -> String
+
+stringify Nil        = "null"
+
+stringify (B True)   = "true"
+stringify (B False)  = "false"
+
+stringify (S string) =
+    wrapQuotes
+  . escape '\\' -- Otherwise parse yields invalid/different strings. [1]
```

Comment:
This doesn't escape control characters like `'\1'` as JSON spec requires. Also, I think you could've instead written something like

```haskell
escape :: [Char] -> String -> String
escape cEscapes s =
  (\ c -> if c `elem` cEscapes then ['\\', '\\', c] else [c]) =<< s
```

to do escaping in one pass by writing `escape ['\\', '\"']`. Anyway, it won't work with control characters or newlines so it doesn't matter.